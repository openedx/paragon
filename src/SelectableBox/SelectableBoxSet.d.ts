import React from 'react';

export interface SelectableBoxSetProps {
  name: string;
  children?: React.ReactNode;
  onChange?: React.ChangeEventHandler<HTMLInputElement>;
  value?: string | number | Array<string | number>;
  defaultValue?: string | number;
  type?: 'radio' | 'checkbox';
  columns?: number;
  className?: string;
}

declare const SelectableBoxSet: React.ForwardRefExoticComponent<SelectableBoxSetProps>;

export default SelectableBoxSet;
