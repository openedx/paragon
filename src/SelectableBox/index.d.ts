import React from 'react';
import { SelectableBoxSetProps } from './SelectableBoxSet';

export interface SelectableBoxProps extends React.HTMLAttributes<HTMLDivElement> {
    value?: string | number;
    checked?: boolean;
    type?: 'radio' | 'checkbox';
    onClick?: (inputRef: React.HTMLInputElement) => void;
    onFocus?: () => void;
    inputHidden?: boolean;
    isIndeterminate?: boolean;
    isInvalid?: boolean;
    className?: string;
}

export type SelectableBoxRef = React.HTMLDivElement & {
    inputRef?: React.HTMLInputElement;
};

declare const SelectableBox: React.ForwardRefExoticComponent<SelectableBoxProps & React.RefAttributes<SelectableBoxRef>> & {
    Set: React.FC<SelectableBoxSetProps>;
};

export default SelectableBox;
