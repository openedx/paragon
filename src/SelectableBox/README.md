---
title: 'SelectableBox'
type: 'component'
components:
- SelectableBox
- SelectableBoxSet
categories:
- Forms
- Content
status: 'New'
designStatus: 'Done'
devStatus: 'In progress'
notes: |
---

A box that has selection states. It can be used as an alternative to a radio button or checkbox set.

The ``SelectableBox`` can contain any kind of content as long as it is not clickable. In other words, there should be no clickable targets distinct from selection.

## Basic Usage

As ``Checkbox``

```jsx live
() => {
  const type = 'checkbox';
  const allCheeseOptions = ['swiss', 'cheddar', 'pepperjack'];
  const [checkedCheeses, { add, remove, set, clear }] = useCheckboxSetValues(['swiss']);

  const handleChange = e => {
    e.target.checked ? add(e.target.value) : remove(e.target.value);
  };
  
  const isInvalid = () => checkedCheeses.includes('swiss');
  const isExtraSmall = useMediaQuery({ maxWidth: breakpoints.extraSmall.maxWidth });
  
  return (
    <div className="bg-light-200 p-3">
      <SelectableBox.Set
        value={checkedCheeses}
        type={type}
        onChange={handleChange}
        name="cheeses"
        columns={isExtraSmall ? 1 : 2}
      >
        <SelectableBox value="swiss" type={type} aria-label="swiss checkbox">
          <div>
            <h3>It is my first SelectableBox</h3>
            <p>Swiss</p>
          </div>
        </SelectableBox>
        <SelectableBox value="cheddar" inputHidden={false} type={type} aria-label="cheddar checkbox">
          Cheddar
        </SelectableBox>
        <SelectableBox
          value="pepperjack"
          inputHidden={false}
          type={type}
          isInvalid={isInvalid()}
          aria-label="pepperjack checkbox"
        >
          <h3>Pepperjack</h3>
        </SelectableBox>
      </SelectableBox.Set>
    </div>
  );
}
```

## As Radio

```jsx live
() => {
  const type = 'radio';
  const [value, setValue] = useState('green');
  const handleChange = e => setValue(e.target.value);
  const isExtraSmall = useMediaQuery({ maxWidth: breakpoints.extraSmall.maxWidth });

  return (
    <SelectableBox.Set
      type={type}
      value={value}
      onChange={handleChange}
      name="colors"
      columns={isExtraSmall ? 1 : 3}
    >
      <SelectableBox value="red" type={type} aria-label="red checkbox">
        <div>
          <h3>It is Red color</h3>
          <p>Select me</p>
        </div>
      </SelectableBox>
      <SelectableBox value="green" inputHidden={false} type={type} aria-label="green-checkbox">
        <h3>Green</h3>
        <p>Leaves and grass</p>
      </SelectableBox>
      <SelectableBox value="blue" inputHidden={false} type={type} aria-label="blue checkbox">
        <h3>Blue</h3>
        <p>The sky</p>
      </SelectableBox>
    </SelectableBox.Set>
  );
}
```
## As Checkbox
As ``Checkbox`` with ``isIndeterminate``

```jsx live
() => {
  const type = 'checkbox';
  const allCheeseOptions = ['swiss', 'cheddar', 'pepperjack'];
  const [checkedCheeses, { add, remove, set, clear }] = useCheckboxSetValues(['swiss']);

  const allChecked = allCheeseOptions.every(value => checkedCheeses.includes(value));
  const someChecked = allCheeseOptions.some(value => checkedCheeses.includes(value));
  const isIndeterminate = someChecked && !allChecked;
  const isExtraSmall = useMediaQuery({ maxWidth: breakpoints.small.maxWidth });

  const handleChange = e => {
    e.target.checked ? add(e.target.value) : remove(e.target.value);
  };

  const handleCheckAllChange = ({ checked }) => {
    checked ? set(allCheeseOptions) : clear();
  };

  return (
    <>
      <div className="mb-3">
        <SelectableBox
          checked={allChecked}
          isIndeterminate={isIndeterminate}
          onClick={handleCheckAllChange}
          inputHidden={false}
          type={type}
          aria-label="all options checkbox"
        >
          All the cheese
        </SelectableBox>
      </div>
      <SelectableBox.Set
        value={checkedCheeses}
        type={type}
        onChange={handleChange}
        name="cheeses"
        columns={isExtraSmall ? 1 : 3}
      >
        <SelectableBox value="swiss" type={type} aria-label="swiss checkbox">
          <div>
            <h3>It is my first SelectableBox</h3>
            <p>Swiss</p>
          </div>
        </SelectableBox>
        <SelectableBox value="cheddar" inputHidden={false} type={type} aria-label="cheddar checkbox">
          Cheddar
        </SelectableBox>
        <SelectableBox value="pepperjack" inputHidden={false} type={type} aria-label="pepperjack checkbox">
          <h3>Pepperjack</h3>
        </SelectableBox>
      </SelectableBox.Set>
    </>
  );
}
```
