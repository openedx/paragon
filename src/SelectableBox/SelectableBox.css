/**
 * Do not edit directly
 * Generated on Wed, 01 Mar 2023 00:27:20 GMT
 */

:root {
  --pgn-spacing-selectable-box-box-space: .75rem;
  --pgn-spacing-selectable-box-border-radius: .25rem;
  --pgn-spacing-selectable-box-padding: 1rem;
}
