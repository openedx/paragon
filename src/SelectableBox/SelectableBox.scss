@import "variables";

.pgn__selectable_box-set {
  display: grid;
  grid-auto-rows: 1fr;
  grid-gap: $selectable-box-space;

  @for $i from $min-cols-number through $max-cols-number {
    &.pgn__selectable_box-set--#{$i} {
      grid-template-columns: repeat(#{$i}, 1fr);
    }
  }

  & > * + * {
    margin: 0;
  }
}

.pgn__selectable_box {
  position: relative;
  height: 100%;
  padding: $selectable-box-padding;
  box-shadow: $level-1-box-shadow;
  border-radius: $selectable-box-border-radius;
  text-align: start;
  background: $white;

  &:focus-visible {
    outline: 1px solid $primary-700;
  }

  .pgn__form-radio,
  .pgn__form-checkbox {
    position: absolute;
    top: $selectable-box-padding;
    inset-inline-end: $selectable-box-padding;

    input {
      margin-inline-end: 0;
    }
  }

  * {
    pointer-events: none;
  }
}

.pgn__selectable_box-active {
  outline: 2px solid $primary-500;
}

.pgn__selectable_box-invalid {
  outline: 2px solid $danger-300;
}
