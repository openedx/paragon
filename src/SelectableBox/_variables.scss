$selectable-box-padding:                   1rem !default;
$selectable-box-border-radius:             .25rem !default;
$selectable-box-space:                     .75rem !default;
$min-cols-number:                          1 !default;
$max-cols-number:                          12 !default;
