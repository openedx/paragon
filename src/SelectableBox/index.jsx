import React, { useRef, useEffect } from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames';
import SelectableBoxSet from './SelectableBoxSet';
import { useCheckboxSetContext } from '../Form/FormCheckboxSetContext';
import { useRadioSetContext } from '../Form/FormRadioSetContext';
import { getInputType } from './utils';

const INPUT_TYPES = [
  'radio',
  'checkbox',
];

const SelectableBox = React.forwardRef(({
  type,
  value,
  checked,
  children,
  isIndeterminate,
  isInvalid,
  onClick,
  onFocus,
  inputHidden,
  className,
  ...props
}, ref) => {
  const inputType = getInputType('SelectableBox', type);
  const { value: radioValue } = useRadioSetContext();
  const { value: checkboxValues = [] } = useCheckboxSetContext();

  const isChecked = () => {
    switch (type) {
      case 'radio':
        return radioValue === value;
      case 'checkbox':
        return checkboxValues.includes(value);
      default:
        return radioValue === value;
    }
  };

  const inputRef = useRef(null);
  const input = React.createElement(inputType, {
    value,
    checked,
    hidden: inputHidden,
    ref: inputRef,
    tabIndex: -1,
    onChange: () => {},
    ...(type === 'checkbox' ? { ...props, isIndeterminate } : { ...props }),
  }, null);

  useEffect(() => {
    if (onClick && inputRef.current) {
      inputRef.current.onclick = () => onClick(inputRef.current);
    }
  }, [onClick]);

  return (
    <div
      role="button"
      onKeyPress={() => inputRef.current.click()}
      onClick={() => inputRef.current.click()}
      onFocus={onFocus}
      className={classNames('pgn__selectable_box', className, {
        'pgn__selectable_box-active': isChecked() || checked,
        'pgn__selectable_box-invalid': isInvalid,
      })}
      tabIndex={0}
      ref={ref}
      {...props}
    >
      {input}
      {children}
    </div>
  );
});

SelectableBox.propTypes = {
  /** Content of the `SelectableBox`. */
  children: PropTypes.node.isRequired,
  /** A value that is passed to the input tag. */
  value: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
  /** Controls whether `SelectableBox` is checked. */
  checked: PropTypes.bool,
  /** Indicates the input type: checkbox or radio. */
  type: PropTypes.oneOf(INPUT_TYPES),
  /** Function that is called when the `SelectableBox` is clicked. */
  onClick: PropTypes.func,
  /** Function that is called when the `SelectableBox` is focused. */
  onFocus: PropTypes.func,
  /** Controls display of the input (checkbox or radio button) on the `SelectableBox`. */
  inputHidden: PropTypes.bool,
  /** Indicates a state for the 'checkbox' `type` when `SelectableBox` is neither checked nor unchecked. */
  isIndeterminate: PropTypes.bool,
  /** Adds errors styles to the `SelectableBox`. */
  isInvalid: PropTypes.bool,
  /** A class that is appended to the base element. */
  className: PropTypes.string,
};

SelectableBox.defaultProps = {
  value: undefined,
  checked: false,
  type: 'radio',
  onClick: () => {},
  onFocus: () => {},
  inputHidden: true,
  isIndeterminate: false,
  isInvalid: false,
  className: undefined,
};

SelectableBox.Set = SelectableBoxSet;
export default SelectableBox;
