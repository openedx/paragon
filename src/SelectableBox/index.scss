.pgn__selectable_box-set {
  display: grid;
  grid-auto-rows: 1fr;
  grid-gap: var(--pgn-spacing-selectable-box-box-space);

  @for $i from 1 through 12 {
    &.pgn__selectable_box-set--#{$i} {
      grid-template-columns: repeat(#{$i}, 1fr);
    }
  }

  & > * + * {
    margin: 0;
  }
}

.pgn__selectable_box {
  position: relative;
  height: 100%;
  padding: var(--pgn-spacing-selectable-box-padding);
  box-shadow: var(--pgn-elevation-box-shadow-level-1);
  border-radius: var(--pgn-spacing-selectable-box-border-radius);
  text-align: start;
  background: var(--pgn-color-white);

  &:focus-visible {
    outline: 1px solid var(--pgn-color-primary-700);
  }

  .pgn__form-radio,
  .pgn__form-checkbox {
    position: absolute;
    top: var(--pgn-spacing-selectable-box-padding);
    inset-inline-end: var(--pgn-spacing-selectable-box-padding);

    input {
      margin-inline-end: 0;
    }
  }

  * {
    pointer-events: none;
  }
}

.pgn__selectable_box-active {
  outline: 2px solid var(--pgn-color-primary-500);
}

.pgn__selectable_box-invalid {
  outline: 2px solid var(--pgn-color-danger-300);
}
