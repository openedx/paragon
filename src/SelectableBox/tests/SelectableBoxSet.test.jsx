import React from 'react';
import { mount } from 'enzyme';
import renderer from 'react-test-renderer';

import { Form } from '../..';
import SelectableBox from '..';

const checkboxType = 'checkbox';
const checkboxText = (text) => `SelectableCheckbox${text}`;

const radioType = 'radio';
const radioText = (text) => `SelectableRadio${text}`;

function SelectableCheckboxSet(props) {
  return (
    <SelectableBox.Set name={radioType} type={checkboxType} {...props}>
      <SelectableBox value={1} type={checkboxType}>{checkboxText(1)}</SelectableBox>
      <SelectableBox value={2} type={checkboxType}>{checkboxText(2)}</SelectableBox>
      <SelectableBox value={3} type={checkboxType}>{checkboxText(3)}</SelectableBox>
    </SelectableBox.Set>
  );
}

function SelectableRadioSet(props) {
  return (
    <SelectableBox.Set name={radioType} type={radioType} {...props}>
      <SelectableBox value={1} type={radioType}>{radioText(1)}</SelectableBox>
      <SelectableBox value={2} type={radioType}>{radioText(2)}</SelectableBox>
      <SelectableBox value={3} type={radioType}>{radioText(3)}</SelectableBox>
    </SelectableBox.Set>
  );
}

describe('<SelectableBox.Set />', () => {
  describe('correct rendering', () => {
    it('renders without props', () => {
      const tree = renderer.create((<SelectableRadioSet name="testName" />)).toJSON();
      expect(tree).toMatchSnapshot();
    });
    it('correct render when type prop is changed', () => {
      const setWrapper = mount(<SelectableRadioSet name="set" />);
      expect(setWrapper.find(Form.RadioSet).length).toBeGreaterThan(0);
      setWrapper.setProps({ type: 'radio' });
      expect(setWrapper.find(Form.RadioSet).length).toBeGreaterThan(0);
      setWrapper.setProps({ type: 'checkbox' });
      expect(setWrapper.find(Form.CheckboxSet).length).toBeGreaterThan(0);
    });
    it('renders with children', () => {
      const wrapper = mount(
        <SelectableCheckboxSet name="testName">{checkboxText(1)}</SelectableCheckboxSet>,
      );
      expect(wrapper.text()).toContain(checkboxText(1));
    });
    it('renders with on change event', () => {
      const onChangeSpy = jest.fn();
      const wrapper = mount(<SelectableCheckboxSet onChange={onChangeSpy} />);
      wrapper.props().onChange();
      expect(onChangeSpy).toHaveBeenCalledTimes(1);
    });
    it('renders with checkbox type', () => {
      const wrapper = mount(<SelectableCheckboxSet />);
      const selectableBoxSet = wrapper.find(Form.CheckboxSet);
      expect(selectableBoxSet.length).toEqual(1);
    });
    it('renders with radio type if neither checkbox nor radio is passed', () => {
      // Mock the `console.error` is intentional because an invalid `type` prop
      // with type `text` specified for `ForwardRef` expects one of the ['radio','checkbox'] parameters.
      // eslint-disable-next-line no-console
      console.error = jest.fn();
      const wrapper = mount(<SelectableCheckboxSet type="text" />);
      const selectableBoxSet = wrapper.find(Form.RadioSet);
      expect(selectableBoxSet.length).toEqual(1);
      // eslint-disable-next-line no-console
      console.error.mockRestore();
    });
    it('renders with radio type', () => {
      const wrapper = mount(<SelectableRadioSet type={radioType} />);
      const selectableBoxSet = wrapper.find(Form.RadioSet);
      expect(selectableBoxSet.length).toEqual(1);
    });
    it('renders with correct number of columns', () => {
      const columns = 10;
      const wrapper = mount(<SelectableRadioSet columns={columns} />);
      const selectableBoxSet = wrapper.find(Form.RadioSet);
      expect(selectableBoxSet.hasClass(`pgn__selectable_box-set--${columns}`)).toBe(true);
    });
  });
});
