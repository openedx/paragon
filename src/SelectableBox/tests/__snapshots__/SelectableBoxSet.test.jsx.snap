// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SelectableBox.Set /> correct rendering renders without props 1`] = `
<div
  className="pgn__selectable_box-set pgn__selectable_box-set--2 pgn__form-control-set"
  role="radiogroup"
>
  <div
    className="pgn__selectable_box"
    onClick={[Function]}
    onFocus={[Function]}
    onKeyPress={[Function]}
    role="button"
    tabIndex={0}
  >
    <div
      className="pgn__form-radio"
    >
      <input
        className="pgn__form-radio-input"
        defaultChecked={false}
        hidden={true}
        id="form-field1"
        name="testName"
        onChange={[Function]}
        tabIndex={-1}
        type="radio"
        value={1}
      />
      <div>
        <label
          className="pgn__form-label"
          htmlFor="form-field1"
        />
      </div>
    </div>
    SelectableRadio1
  </div>
  <div
    className="pgn__selectable_box"
    onClick={[Function]}
    onFocus={[Function]}
    onKeyPress={[Function]}
    role="button"
    tabIndex={0}
  >
    <div
      className="pgn__form-radio"
    >
      <input
        className="pgn__form-radio-input"
        defaultChecked={false}
        hidden={true}
        id="form-field2"
        name="testName"
        onChange={[Function]}
        tabIndex={-1}
        type="radio"
        value={2}
      />
      <div>
        <label
          className="pgn__form-label"
          htmlFor="form-field2"
        />
      </div>
    </div>
    SelectableRadio2
  </div>
  <div
    className="pgn__selectable_box"
    onClick={[Function]}
    onFocus={[Function]}
    onKeyPress={[Function]}
    role="button"
    tabIndex={0}
  >
    <div
      className="pgn__form-radio"
    >
      <input
        className="pgn__form-radio-input"
        defaultChecked={false}
        hidden={true}
        id="form-field3"
        name="testName"
        onChange={[Function]}
        tabIndex={-1}
        type="radio"
        value={3}
      />
      <div>
        <label
          className="pgn__form-label"
          htmlFor="form-field3"
        />
      </div>
    </div>
    SelectableRadio3
  </div>
</div>
`;
